app/src/org/commcare/dalvik/activities/CreatePinActivity.java

```diff
@@ -0,0 +1,220 @@
+            userRecord.clearPrimedPassword();
+            CommCareApplication._().getCurrentApp().getStorage(UserKeyRecord.class).write(userRecord);
+        }
+
+        setListeners();
+        setInitialEntryMode();
+    }
+
+
+    private void setListeners() {
+        enterPinBox.addTextChangedListener(getPinTextWatcher(continueButton));
+
+        cancelButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                setResult(RESULT_CANCELED);
+                finish();
+            }
+        });
+
+        continueButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                if (!inConfirmMode) {
+                    processInitialPinEntry();
+                } else {
+                    processConfirmPinEntry();
+                }
+            }
+        });
+    }
+
+    private void processInitialPinEntry() {
+        firstRoundPin = enterPinBox.getText().toString();
+        setConfirmMode();
+    }
+
+    private void processConfirmPinEntry() {
+        String enteredPin = enterPinBox.getText().toString();
+        if (enteredPin.equals(firstRoundPin)) {
+            assignPin(enteredPin);
+            setResult(RESULT_OK);
+            finish();
+        } else {
+            Toast.makeText(this, Localization.get("pins.dont.match"), Toast.LENGTH_SHORT).show();
+            setInitialEntryMode();
+        }
+    }
+
+    private void setInitialEntryMode() {
+        enterPinBox.setText("");
+        enterPinBox.requestFocus();
+        continueButton.setText(Localization.get("pin.continue.button"));
+        if (CommCareApplication._().getRecordForCurrentUser().hasPinSet()) {
+            promptText.setText(Localization.get("pin.directive.reset"));
+        } else {
+            promptText.setText(Localization.get("pin.directive.new"));
+        }
+        inConfirmMode = false;
+    }
+
+    private void setConfirmMode() {
+        enterPinBox.setText("");
+        enterPinBox.requestFocus();
+        continueButton.setText(Localization.get("pin.confirm.button"));
+        promptText.setText(Localization.get("pin.directive.confirm"));
+        inConfirmMode = true;
+    }
+
+    private void assignPin(String pin) {
+        userRecord.assignPinToRecord(pin, unhashedUserPassword);
+        CommCareApplication._().getCurrentApp().getStorage(UserKeyRecord.class).write(userRecord);
+    }
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        super.onCreateOptionsMenu(menu);
+        menu.add(0, MENU_REMEMBER_PW_AND_LOGOUT, 0,
+                Localization.get("remember.password.for.next.login"));
+        return true;
+    }
+
+    @Override
+    public boolean onOptionsItemSelected(MenuItem item) {
+        if (item.getItemId() == MENU_REMEMBER_PW_AND_LOGOUT) {
+            launchRememberPasswordConfirmDialog();
+        }
+        return true;
```

Comment:
Saying that because a lot of other code seems to do that; not sure exactly what the implications